source/memory-store.ts

```diff
@@ -30,19 +36,37 @@ export default class MemoryStore implements Store {
 	windowMs!: number
 
 	/**
-	 * The map that stores the number of hits for each client in memory.
+	 * These two maps store usage (requests) and reset time by key (for example, IP
+	 * addresses or API keys).
+	 *
+	 * They are split into two to avoid having to iterate through the entire set to
+	 * determine which ones need reset. Instead, `Client`s are moved from `previous`
+	 * to `current` as they hit the endpoint. Once `windowMs` has elapsed, all clients
+	 * left in `previous`, i.e., those that have not made any recent requests, are
+	 * known to be expired. At that point, the cache pool is filled from `previous`,
+	 * and any remaining `Client`s are cleared from memory.
 	 */
-	hits!: {
-		[key: string]: number | undefined
-	}
+	previous = new Map<string, Client>()
+	current = new Map<string, Client>()
+
+	/**
+	 * The cache of unused clients, kept to reduce the number of objects created
+	 * and destroyed. Improves performance, at the expense of a small amount of RAM.
+	 *
+	 * Each entry takes up 152 bytes. In one benchmark, the total time taken to handle
+	 * 100M requests was reduced from 70.184s to 47.862s (~32% improvement) with
+	 * ~5.022 MB extra RAM used.
+	 */
+	pool: Client[] = []
 
 	/**
-	 * The time at which all hit counts will be reset.
+	 * Used to calculate how many entries to keep in the pool.
 	 */
-	resetTime!: Date
+	numberOfCreatedClients = 0
+	recentClientsCount: number[] = []
```

Comment:
Hum.. I concede that the old names weren't great, but this doesn't feel right either.

The first one is the number of keys that we've seen during the current window that we didn't see during the previous window. Ideally pool should begin the window with exactly this many Client objects, so that 0 are created.

The second one is a series of previous values for the first. So it should probably just be the plural form of the name of the first one. 

Some ideas: 
* `idealPoolSize` & `idealPoolSizes`?
* `currentNewKeys` & `previousNewKeys`
* `uncachedCount` & `uncachedCounts`

I don't know. Naming things is hard.

Reply:
`uncachedCount` & `uncachedCounts` sounds good 👍🏾

Reply:
done in d89e514